odoo/addons/base/models/res_lang.py

```diff
@@ -155,7 +169,190 @@ def fix_datetime_format(format):
+                                       version),
+                                       lang + ".tar.xz")
+
+    @api.model
+    def _extract_to_i18n_folder(self, addons, tmp, extracted_file):
+        path = http.addons_manifest[addons]['addons_path']
+        if path not in ad_paths:
+            _logger.info("Skip translations extraction for module %s, path %s not in addons-path", addons, path)
+            return False
+
+        if not os.access(path, os.W_OK):
+            _logger.warning("Path %s is not writeable", path)
+            return False
+
+        src_file = os.path.join(tmp, extracted_file)
+        dst_file = os.path.join(path, extracted_file)
+        shutil.move(src_file, dst_file)
+
+    @api.model
+    def _extract_to_attachment(self, addons, tmp, extracted_file):
+        module = self.env['ir.module.module'].search([('name', '=', addons)], limit=1)
+        if not module:
+            _logger.info("Skip translations extraction for unexpected module %s", addons)
+            return
+        if module.state != 'installed':
+            _logger.info("Skip translations extraction, module %s is in state %s", addons, module.state)
+            return
+
+        src_file = os.path.join(tmp, extracted_file)
+        with open(src_file, 'r') as po_file:
+            po_content = base64.b64encode(po_file.read().encode()).decode()
+
+        # sudo to be able to search and write, even if no rights on ir.module.module
+        Attachment = self.env['ir.attachment'].sudo()
+
+        attached_po = Attachment.search([
+            ('res_model', '=', 'ir.module.module'),
+            ('res_id', '=', module.id)
+            ('name', '=', extracted_file)
+        ], limit=1)
+        if attached_po:
+            attached_po.write({'datas': po_content})
+        else:
+            attached_po = Attachment.create({
+                'datas': po_content,
+                'res_model': 'ir.module.module',
+                'res_id': module.id,
+                'name': extracted_file,
+            })
+        return attached_po
+
+    @api.model
+    def _extract_i18n_file_content(self, fileobj, lang, module_list, extraction_method):
+        """ Extract the translations from the given archive
+
+        The expected archive is a .tar.xz file using the structure:
+            <module>/i18n/<lang>.po
+
+        Regional variants are accepted (e.g. fr.tar.xz can contain both fr and
+        fr_BE files)
+
+        :param fileobj: a file object containing the compressed translation files
+        :param lang: the simplified language code to load (e.g. 'fr')
+        :param module_list: the list of modules to update from this archive
+        :param extraction_method: the method to call with the extracted .po file
+        """
+        with tarfile.open(mode='r:xz', fileobj=fileobj) as tar_content:
+            with tempfile.TemporaryDirectory() as tmp:
+                for filename in tar_content.getnames():
```

Comment:
might be worth iterating on the TarFile directly to get TarEntry objects so we can check the entry's size, & keep a running tally or somesuch to avoid zip bombs?